### ISSUE TITLE
[BUG] Log auto-scaling not showing all data

### ISSUE THREAD

**Issue by EfremBraun** [bokeh/bokeh#11508]
#### ALL software version info (bokeh, python, notebook, OS, browser, any other relevant packages)
Bokeh 2.3.3
Python 3.8.5
(Also tested with Python 3.9.6 and Bokeh 2.4.0dev3-7-gf312f679f)
MacOs Catalina 10.15.7
Google Chrome 92.0.4515.131

#### Description of expected behavior and the observed behavior
When log-scaling is used on a data set I'm working with, the auto-scaling does not show all the data. I thought this might have been the same issue as #11378, but I checked out the Bokeh 2.4 branch after the #11379 pull request and the issue is still present.

#### Complete, minimal, self-contained example code that reproduces the issue

```
import numpy as np
from bokeh.plotting import figure
from bokeh.io import show

rawdata = np.loadtxt('data.txt')
x = rawdata[:,0]
y = rawdata[:,1]

fig = figure(y_axis_type='log')
fig.line(x=x, y=y)
show(fig)
```
This is the data set:
[data.txt](https://github.com/bokeh/bokeh/files/6984997/data.txt)

#### Screenshots or screencasts of the bug in action
This is what appears:
<img width="609" alt="Screen Shot 2021-08-13 at 4 10 57 PM" src="https://user-images.githubusercontent.com/6618019/129413997-18f3d82a-8256-42f8-89c8-539fc28ab0ea.png">
But this is all the data (shown by manually zooming out):
<img width="598" alt="Screen Shot 2021-08-13 at 4 11 33 PM" src="https://user-images.githubusercontent.com/6618019/129414046-62c3a4e7-99ca-4094-a150-59ecd284a054.png">

**Comment by mattpap** [comment]
In JS console:
```
data_range1d.js:98 [bokeh] could not determine minimum data value for log axis, DataRange1d using value 3.285534485589579e+32
```

**Comment by bryevdv** [comment]
Codepath is here:

https://github.com/bokeh/bokeh/blob/0b45558db912a6e9a73e6116aeaa6c0ae4d2ae65/bokehjs/src/lib/models/ranges/data_range1d.ts#L167-L173

**Comment by bryevdv** [comment]
@EfremBraun It looks like there are 0 y-values? If so, this is `working as intended`

**Comment by EfremBraun** [comment]
I didn't think the 0s are the issue...

Copying the example given at https://docs.bokeh.org/en/latest/docs/user_guide/plotting.html#log-scale-axes and slightly modifying it to include 0s:
```
from bokeh.plotting import figure, output_file, show

x = [0.1, 0.5, 1.0, 1.5, 2.0, 2.5, 3.0]
y = [1E5, 0, 1E-2, 0, 3E3, 4E1, 8E21]

output_file("log.html")

# create a new plot with a log axis type
p = figure(plot_width=400, plot_height=400, y_axis_type="log")

p.line(x, y, line_width=2)
p.circle(x, y, fill_color="white", size=8)

show(p)
```

The result is a very nice looking plot:
<img width="402" alt="Screen Shot 2021-08-13 at 5 23 14 PM" src="https://user-images.githubusercontent.com/6618019/129419772-ddfc959a-b4eb-4f37-ad88-413c567d91cb.png">

**Comment by EfremBraun** [comment]
You are indeed correct though that the first plot is showing the min as 1/100 as the max, so the code must be going into the `if` statement you showed above. But why would that dataset go into that `if` statement, whereas the example I just posted shows a correct min value that disregards the 0 values?

**Comment by EfremBraun** [comment]
Here's a much smaller data set that still illustrates the issue (maybe will help with troubleshooting):
[data.txt](https://github.com/bokeh/bokeh/files/6985275/data.txt)

**Comment by EfremBraun** [comment]
Here's a much smaller example that still illustrates the problem (hopefully helpful for troubleshooting):
```
from bokeh.plotting import figure, output_file, show

x = [7.722499966621398926e-01, 7.727500200271606445e-01,
7.732499837875366211e-01, 7.737500071525573730e-01, 7.742500305175781250e-01,
7.747499942779541016e-01, 7.752499580383300781e-01, 7.757500410079956055e-01,
7.762500047683715820e-01, 7.767499685287475586e-01, 7.772499918937683105e-01,
7.777500152587890625e-01, 7.782499790191650391e-01, 7.787500023841857910e-01,
7.857500314712524414e-01, 7.862499952316284180e-01, 7.867499589920043945e-01,
7.872500419616699219e-01, 8.417500257492065430e-01, 8.422499895095825195e-01,
8.427499532699584961e-01, 8.432500362396240234e-01, 8.437500000000000000e-01,
8.442499637603759766e-01, 8.447500467300415039e-01]
y = [0.000000000000000000e+00, 8.164452529434230836e+22,
0.000000000000000000e+00, 0.000000000000000000e+00, 7.314143412752266717e+22,
0.000000000000000000e+00, 6.839288802346721188e+22, 0.000000000000000000e+00,
0.000000000000000000e+00, 6.232344689415452361e+22, 0.000000000000000000e+00,
0.000000000000000000e+00, 0.000000000000000000e+00, 0.000000000000000000e+00,
0.000000000000000000e+00, 0.000000000000000000e+00, 0.000000000000000000e+00,
1.661581512390552584e+21, 1.005171116507131360e+17, 8.466177779596089600e+16,
7.311184945273668800e+16, 6.434035489362382400e+16, 5.745531645071752000e+16,
0.000000000000000000e+00, 4.731234037419803200e+16]

output_file("log.html")

# create a new plot with a log axis type
p = figure(plot_width=400, plot_height=400, y_axis_type="log")

p.line(x, y, line_width=2)
p.circle(x, y, fill_color="white", size=8)

show(p)
```

**Comment by EfremBraun** [comment]
I've isolated the issue. See:
```
from bokeh.plotting import figure, output_file, show

x = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
y = [0.000000000000000000e+00, 8.164452529434230836e+22,
0.000000000000000000e+00, 0.000000000000000000e+00, 7.314143412752266717e+22,
0.000000000000000000e+00, 6.232344689415452361e+22, 0.000000000000000000e+00, # Remove this line (and 3 numbers from x) and the issue goes away
0.000000000000000000e+00, 0.000000000000000000e+00, 0.000000000000000000e+00,
1.661581512390552584e+21, 1.005171116507131360e+17, 8.466177779596089600e+16,
7.311184945273668800e+16, 6.434035489362382400e+16, 5.745531645071752000e+16,
0.000000000000000000e+00, 4.731234037419803200e+16]

output_file("log.html")

# create a new plot with a log axis type
p = figure(plot_width=400, plot_height=400, y_axis_type="log")

p.line(x, y, line_width=2)
p.circle(x, y, fill_color="white", size=8)

show(p)
```

I see nothing different about the data in that one line that seems to cause the issue.

**Comment by EfremBraun** [comment]
This is the result of the code directly above:
<img width="419" alt="Screen Shot 2021-08-13 at 5 50 39 PM" src="https://user-images.githubusercontent.com/6618019/129421717-3d740815-b58e-4e25-b923-f5811cd94b8c.png">

And when that one line of 3 data points is removed, this is the result:
<img width="403" alt="Screen Shot 2021-08-13 at 5 50 44 PM" src="https://user-images.githubusercontent.com/6618019/129421722-e698756e-157a-4ea4-bf1d-fbada673b469.png">

Can't figure out why that difference causes such a big change...

**Comment by bryevdv** [comment]
Well, the min/max compute actually relies on the "plot bounds" which are generated by the individual renderers spatial indexes:
```typescript
    // update the raw data bounds for all renderers we care about
    let total_bounds = this._compute_plot_bounds(renderers, bounds)

    if (ratio != null)
      total_bounds = this.adjust_bounds_for_aspect(total_bounds, ratio)

    this._plot_bounds.set(plot, total_bounds)

    // compute the min/mix for our specified dimension
    const [min, max] = this._compute_min_max(this._plot_bounds.values(), dimension)
```
So then that could be an issue wit the initial naive spatial indexing, or the adjustment made for aspect-preservation. Will definitely require some in-depth investigation.

**Comment by bryevdv** [comment]
Ok well, the proximate cause is that the `update` function is getting called different bounds: 

* first call: `{x0: 1, y0: 0, x1: 19, y1: 8.16445252943423e+22}`
* second call: `{x0: 1, y0: 47312340374198030, x1: 19, y1: 8.16445252943423e+22}`

As you can see the second call (which is what ends up finally rendered) has the non-zero y0 value. That's the "working" case. But this means the issue lies outside `DataRange1d` in the plot's spatial indexing an bounds computation. 

cc @mattpap

**Comment by bryevdv** [comment]
As an aside I forgot that plot spatial indexing was leverage for obtaining the min/max. That means the "next lowest min" ideas is not really feasible. So the fix for this issue will simply be to make thing consistent and always apply the expected default bounds. What we could easily make configurable is the number of decades to scale down, right now it is hard-coded to 100. but I am not really sure that is worth doing.

**Comment by EfremBraun** [comment]
Works great now. Thanks!

**Comment by bryevdv** [comment]
Just noting that the behavior in this PR was slightly wrong to begin with and will change a bit after https://github.com/bokeh/bokeh/pull/14340 is merged for 3.7. Details are available there but TLDR in the image below the isolated dots that are present, but not visible in lines due to adjacent 0 or nan values, were not being taken into account. But all data should factor in, and this will be the case going forward.

![Image](https://github.com/user-attachments/assets/34159fb7-47df-4442-9c2e-6537c715e7ca)
